rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,124 @@
+
+Because only a single version of the client tools will be cached at a time, we do not need to worry about cleaning up the cache.
+
+### Package upgrade
+
+As an alternative to caching, we've considered directly upgrading the Teleport packages.
+
+This approach will be difficult to implement because we support multiple package repositories and installation methods. The amount of maintenance and testing required to support this solution is probably not worth the effort at this time.
+
+Another downside to this approach would be that upgrades would require sudo/admin privileges.
+
+### Config
+
+Automatic updates for client tools can be configured through tsh configuration. If `DISABLE_AUTO_UPDATE=true`, then auto updates for client tools will be disabled. If `DISABLE_UPDATE_PROMPT=true`, then the user will not be prompted to confirm the update. Both values will default to false.
+
+### Binary Execution
+
+The client tools will need a way to execute the cached binaries. When tsh/tctl is executed, it will now need to decide whether to execute the command as usual, or fork a new process using the cached binary.
+
+One way to handle execution of cached binaries is to use the os.Executable function. os.Executable give us the path of the binary executable. The flow might look something like:
+- If executing `tsh login` execute as usual (cached tsh binary will not support TouchID on macOS)
+- If executable is in cache directory (.tsh/bin/tsh), then execute command as usual
+- Else try to execute tsh from cache if exists
+
+An alternative method could be use to a flag like `--no-cache` to determine if a cached version should be executed. The flow might look something like:
+- If `--no-cache=true`, execute as usual
+- If `--no-cache=false`, try to execute tsh from cache with `--no-cache=true`
+- If tsh does not exist in cache, execute as usual
+
+Another option could be to have users update their `PATH` env with `export PATH=$HOME/.tsh/bin:$PATH`. Users would then execute the cached binary directly. This seems like a simpler approach then the above methods. The problem is that this requries an extra step for the users, and the cached binary would not support TouchID for macOS users.
+
+## UX
+
+Ideally, we'd like this feature to be integrated seamlessly. Users of the Teleport client tools should not need to search for documentation and spend time figuring out how to enable auto updates for their cleint tools.
+
+After client tools auto updates is deployed and the first time the client tools are executed, the user will be asked two questions: 1) if they would like to update to the latest version (default yes) and 2) if they would like to enroll in auto updates for client tools (default yes).
```

Comment:
Would it make sense to turn this on by default when Cloud clusters are targeted, and require an opt-out command to disable it?

Reply:
I don't have a strong opinion on either way. I think most users would opt-in to auto updates so I think it'd be fine to turn this on by default.

It could also be nice to prompt the user when auto updates are first available to let the user know that auto updates is configurable.

Reply:
I don't think we should ask whether they want to enroll or not. Let's just enable it by default.